Removed unnecessary DB queries when editing/validating RSQL query in …

Sergey-G-dev opened the pull request:
Based on feedback from (#1023)

Added valid TargetFields to RSQL validation when editing Target Filter.

Validation now also considers allowed Target Fields when validating RSQL filters.

bogdan-bondar commented on hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtility.java:
```diff
@@ -161,6 +167,109 @@ private static Node parseRsql(final String rsql) {
     }
 
+
+    private static final class ValidationRSQLVisitor<A extends Enum<A> & FieldNameProvider> implements RSQLVisitor<Boolean, String>{
```
As you've already mentioned just duplicating code is not a great idea (DRY principle, etc.). So my suggestion would be to either create and AbstractFieldNameRSQLVisitor class that exposes helper methods such as #getFieldEnumByName, #getSubAttributesFrom, etc. or alternatively use the composition by injecting some helper class to whom you could delegate the corresponding parse/validate operations

Sergey-G-dev replied:
I've created an abstract superclass where I moved all commonly used code by both RSQL visitors. The visitors now extend that abstract class.

bogdan-bondar commented on hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtility.java:
```diff
@@ -161,6 +167,109 @@ private static Node parseRsql(final String rsql) {
+
+            // sub entity need minium 1 dot
+            if (!propertyEnum.getSubEntityAttributes().isEmpty() && graph.length < 2) {
+                throw new RSQLParameterUnsupportedFieldException();
```
I wouldn't just throw the plain RSQLParameterUnsupportedFieldException, because in that case user will not be able to understand where the problem occurred, using #createRSQLParameterUnsupportedException gives more information, however you could also improve upon that

Sergey-G-dev replied:
Done! Using #createRSQLParameterUnsupportedException method now.

bogdan-bondar commented on hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtilityTest.java:
```diff
@@ -99,6 +100,43 @@ TenantConfigurationManagementHolder tenantConfigurationManagementHolder() {
     private static final TenantConfigurationValue<String> TEST_POLLING_OVERDUE_TIME_INTERVAL = TenantConfigurationValue
             .<String> builder().value("00:07:37").build();
 
+    @Test(expected = RSQLParameterUnsupportedFieldException.class)
```
I wouldn't define the expected exception like that: first of all it is not consistent with the rest of the test class and secondly it is deprecated with Junit 5, so it would make future migration more involved. I would suggest either following the used pattern of try/catch/fail or use the AssertJ with its assertThatExceptionOfType(exceptionType).isThrownBy(throwingCallable)

Sergey-G-dev replied:
I decided to go ahead with the functional style using AssertJ since it is less code and hence more readable.

bogdan-bondar commented on hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtilityTest.java:
```diff
@@ -99,6 +100,43 @@ TenantConfigurationManagementHolder tenantConfigurationManagementHolder() {
+    @Test(expected = RSQLParameterUnsupportedFieldException.class)
+    @Description("Testing exception in case of not allowed subkey")
+    public void rsqlUnsupportedSubkeyThrowException(){
+        final String rsql = "ASSIGNEDDS.unsupported == abcd and ASSIGNEDDS.version == 0123";
```
I would not stick to the Target fields in this test class as it should be generic. However it makes totally sense to test different behaviour of #isValid method by extending the TestFieldEnum defined at the bottom

Sergey-G-dev replied:
Using TestFieldEnum now instead of TargetFields as the test object. TestFieldEnum adjusted accordingly.

bogdan-bondar commented on hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtilityTest.java:
```diff
@@ -99,6 +100,43 @@ TenantConfigurationManagementHolder tenantConfigurationManagementHolder() {
+    @Test
+    @Description("Testing allowed RSQL keys based on TargetFields.class")
+    public void rsqlValidTargetFields(){
+        final String rsql1 = "ID == '0123' and NAME == abcd and DESCRIPTION == absd" +
```
In case you would really like to test the Target fields explicitly you could extend the RSQLTargetFieldTest or introduce a new one e.g. RSQLTargetFieldValidationTest

Sergey-G-dev replied:
Ended up creating a new RSQLTargetFieldValidationTest class since it doesn't really fit into the existing RSQLTargetFieldTest, which operates heavily on JPA/Repository layer.

bogdan-bondar commented on hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtility.java:
```diff
@@ -161,6 +167,109 @@ private static Node parseRsql(final String rsql) {
+        @Override
+        public Boolean visit(OrNode node, String param) {
+            for(Node child: node.getChildren()){
+                if(child.accept(this, param))
```
I'm not sure if the boolean algebra semantic is valid here... I guess AndNode and OrNode should behave same here, because you are checking the validity of the expression, not the condition evaluation. Could you add some tests to verify it?

Sergey-G-dev replied:
Good catch, you are absolutely correct here! Fixed.

bogdan-bondar reviewed (changes requested):
Thanks for adaptation, see the comments above ;)

schabdo commented:
@hawkbit-bot verify please

Sergey-G-dev commented on hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtility.java:
```diff
@@ -139,14 +140,19 @@ private RSQLUtility() {
      */
-    public static void isValid(final String rsql) {
-        parseRsql(rsql);
+    public static boolean isValid(final String rsql){
```
Makes sense. Done!

Sergey-G-dev commented:
Please kindly review the changes where I addressed all the issues from the previous commit pointed out by @bogdan-bondar. See the comments to each issue above.

schabdo commented:
@hawkbit-bot verify please

schabdo commented on hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLTargetFieldValidationTest.java:
```diff
@@ -0,0 +1,40 @@
+
+import static org.junit.Assert.assertTrue;
+
+public class RSQLTargetFieldValidationTest {
```
My 5 cents to this PR: Please don't forget to add the Allure annotations `@Feature(...)` and
`@Story(....)`	to the test class :)

bogdan-bondar commented on hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLTargetFieldValidationTest.java:
```diff
@@ -0,0 +1,40 @@
+public class RSQLTargetFieldValidationTest {
+    @Test
+    @Description("Testing allowed RSQL keys based on TargetFields.class")
+    public void rsqlValidTargetFields(){
```
we should also test some invalid queries in a separate test method(s)

bogdan-bondar commented on hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtilityTest.java:
```diff
@@ -99,6 +104,46 @@ TenantConfigurationManagementHolder tenantConfigurationManagementHolder() {
+    @Description("Testing throwing exception in case of not allowed RSQL key")
+    public void rsqlUnsupportedFieldExceptionTest(){
+        final String rsql1 = "wrongfield == abcd";
+        assertThatExceptionOfType(RSQLParameterUnsupportedFieldException.class)
```
It is necessary to test the messages of `RSQLParameterUnsupportedFieldException` as well as `RSQLParserException`, probably in a separate method

stefbehl commented:
Hi @Sergey-G-dev, we would like to merge your pull request soon (because of the Spring migration, we were not able to merge it end of last year). Would be great if you could take another look at the Sonar finding and the review comments from @bogdan-bondar. Thanks!

krishna-devolo commented:
Hi, 

I am Krishna, a new employee at devolo who will take over Sergey's pull requests about hawkBit.  

From what I understand, the issue behind this pull request is solved by pull request [#1121](https://github.com/eclipse/hawkbit/pull/1121). This pull request can be closed, unless I am missing something?

stefbehl commented:
Hi @krishna-devolo, yes that's correct. As we did not hear back from you, we had to create a new one (which is however completely based on Sergey's PR). The PR is about to be merged in the coming days. If you have any questions or some last-minute feedback on the PR, feel free to comment on the PR directly. Thanks!

floruschbaschan commented:
Hi @krishna-devolo, glad to hear that you will take over Sergey's PR(s). Could you please have a look at #1046? I think there are some minor left-overs.

stefbehl commented:
Closing this pull request as issue #1023 has been solved by pull request #1121.
